src/NCalc.Core/Parser/LogicalExpressionParser.cs

```diff
@@ -364,12 +364,19 @@ private static Parser<LogicalExpression> CreateExpressionParser(CultureInfo cult
 
         var guid = OneOf(guidWithHyphens, guidWithoutHyphens);
 
-        // primary => GUID | NUMBER | identifier| DateTime | string | function | boolean | groupExpression | list ;
+        var integralNumber = Select<LogicalExpressionParserContext, LogicalExpression>((ctx) =>
+        {
+            if (ctx.Options.HasFlag(ExpressionOptions.LongAsDefault))
+                return longNumber;
+
+            return OneOf(intNumber, longNumber);
+        });
+
+        // primary => GUID | NUMBER | identifier | DateTime | string | function | boolean | groupExpression | list ;
         var primary = OneOf(
             guid,
             hexOctBinNumber,
-            intNumber,
-            longNumber,
+            integralNumber,
```

Comment:
`Select` is not `ISeekable` that could prevent the whole `primary` parser from using a lookup and kill perf (or at least the nice free optimization parlot will find out).

You can try to use `Lookup` after the Select to give the list of valid chars since the expected chars is known.

`Select(...).Lookup(integerNumber, longNumber)`

Reply:
The other option (which I think I already suggested) is to only parse `Number<long>` but use `Then()` to convert to `int` based on the option of now. One line, much simpler, no Select